### PR TITLE
hashes/sha256: initial commit for hmac-sha256 computation

### PR DESCRIPTION
**Pull request by BytesGalore**
added initial version of `hmac-sha256` computation function.

### CODE REVIEW

File: sys/crypto/sha256.c [hunk 1]
```diff
@@ -259,3 +259,61 @@ unsigned char *sha256(const unsigned char *d, size_t n, unsigned char *md)
...
+                            const unsigned *message,
+                            size_t message_length,
+                            unsigned char *result)
+{
+    unsigned char k[64];
+    memset((void *)k, 0x00, 64);
+
+    if (key_length > 64) {
+        sha256(key, key_length, k);
+    } else {
```
Comment (OlegHahm) [inline sys/crypto/sha256.c]: Linebreak after `}`

**Comment by OlegHahm** [comment]
Did you implement from scratch?

**Comment by OlegHahm** [comment]
Needs a rebase

File: sys/crypto/sha256.c (hunk 1 above)
Comment (BytesGalore) [inline sys/crypto/sha256.c]: done.

**Comment by BytesGalore** [comment]
yes, from scratch.
rebased.

**Comment by OlegHahm** [comment]
Have you compared your implementation with the one proposed in [RFC4634](https://tools.ietf.org/html/rfc4634#section-8.3)?

I would also vote for adding more test vectors, maybe the ones from [RFC4868](https://tools.ietf.org/html/rfc4868#section-2.7).

**Comment by BytesGalore** [comment]
I had a look in [RFC4634](https://tools.ietf.org/html/rfc4634#section-8.3) but have not compared it yet.
My implementation follows roughly a pseudocode similar the one found on wikipedia [1]

[RFC4868](https://tools.ietf.org/html/rfc4868#section-2.7) is a good source for testvectors (I have been too lazy to add them :blush:)

[1] https://en.wikipedia.org/wiki/Hash-based_message_authentication_code

**Comment by OlegHahm** [comment]
Are you planning to add the additional test vectors? Particular for self-implemented security related stuff, I would like to have some more checks.

File: sys/include/crypto/sha256.h [hunk 2]
```diff
@@ -100,6 +100,23 @@ void sha256_final(unsigned char digest[32], sha256_context_t *ctx);
  */
 unsigned char *sha256(const unsigned char *d, size_t n, unsigned char *md);
 
+/**
+ * @brief function to compute a hmac-sha256 from a given message
+ *
+ * @param[in] key key used in the hmac-sha256 computation
+ * @param[in] key_length the size in bytes of the key
+ * @param[in] message poiter to the message to generate the hmac-sha256
```
Comment (A-Paul) [inline sys/include/crypto/sha256.h]: pointer :)

File: sys/include/crypto/sha256.h [hunk 3]
```diff
@@ -100,6 +100,23 @@ void sha256_final(unsigned char digest[32], sha256_context_t *ctx);
...
+ * @param[in] message_length the length of the message in bytes
+ * @param[out] result the computed hmac-sha256,
+ *             length MUST be SHA256_DIGEST_LENGTH
+ *             if result == NULL, a static buffer is used
+ */
+unsigned char * hmac_sha256(const unsigned char *key,
+                            size_t key_length,
+                            const unsigned *message,
```
Comment (A-Paul) [inline sys/include/crypto/sha256.h]: Defaults to `unsigned int`. On purpose?

Comment (BytesGalore) [inline sys/include/crypto/sha256.h]: Its somehow intended.
I want to keep it consistent with the undocumented behaviour of the `sha256` function.
But now that you pointed me I'm not really happy with this thread-unsafe behaviour, neither in `sha256` nor here.

**Comment by BytesGalore** [comment]
@OlegHahm added further tests with testvectors form [RFC4868#Section-2.7.1](https://tools.ietf.org/html/rfc4868#section-2.7.1)
(Though I cannot test it form here. Just used the GITHub webeditor for the changes)

File: tests/unittests/tests-crypto/tests-crypto-sha256.c [hunk 4]
```diff
@@ -81,10 +81,105 @@ static void test_crypto_sha256_hash_sequence(void)
...
+                 "b0344c61d8db38535ca8afceaf0bf12b881dc200c9833da726e9376c2e32cff7", hmac));
+
+    /* Test Case PRF-2: */
+    const char strPRF2[] = "what do ya want for nothing?";
+    /* clear the key (we used 20 bytes so we clear only 20) */
+    memset(key, 0x0, 20);
+    key[0] = 'J'
+    key[1] = 'e'
+    key[2] = 'f'
+    key[3] = 'e'
```
Comment (OlegHahm) [inline tests/unittests/tests-crypto/tests-crypto-sha256.c]: `;` are missing in the four lines above.

**Comment by OlegHahm** [comment]
```
crypto_sha256_tests.test_crypto_hmac_sha256_hash_sequence (tests/unittests/tests-crypto/tests-crypto-sha256.c 141) compare_str_vs_digest( "773ea91e36800e46854db8ebd09181a72959098b3ef8c122d9635514ced565fe", hmac)
...................
run 21 failures 1
```

**Comment by BytesGalore** [comment]
@OlegHahm fixed both (hopefully).
I mistakenly used `strlen()` to determine the size of the data to be hashed for this test [1] and also the following one [2].
Since they are obviously no `\0` terminated strings it didn't work.

[1] https://github.com/RIOT-OS/RIOT/pull/4679/files#diff-763818222f6f80b33045cbf067bcb3e9R139
[2] https://github.com/RIOT-OS/RIOT/pull/4679/files#diff-763818222f6f80b33045cbf067bcb3e9R155

**Comment by OlegHahm** [comment]
Hm, no, the test is still failing.

**Comment by BytesGalore** [comment]
I initialized the bytes of `key` for 3/5 and `strPRF3` with wrong values.
`20 * 0xa` vs `20 * 0xaa` and `50 * 0xd` vs `50 * 0xdd`

**Comment by OlegHahm** [comment]
This test is now succeeding, but:
```
crypto_sha256_tests.test_crypto_hmac_sha256_hash_sequence (tests/unittests/tests-crypto/tests-crypto-sha256.c 166) compare_str_vs_digest( "60e431591ee0b67f0d8a26aacbf5b77f8e0bc6213728c5140546040f0ee37f54", hmac)
```

**Comment by BytesGalore** [comment]
ok, I will have a look this afternoon.
I guess something goes wrong hashing the `longKey` [1].
Thx for the testing and updating me :)

[1] https://github.com/RIOT-OS/RIOT/pull/4679/files#diff-1905d563e1e656f06f81fb3b030b5dcdR273

**Comment by OlegHahm** [comment]
Rebase is required because of #4715 (moved sha256 to hashes).

**Comment by BytesGalore** [comment]
yup just saw it.

**Comment by BytesGalore** [comment]
Reabsed and fixed the last error.

I failed to initialize `longKey[]` properly. 
But now its done and the unittests pass :D .

File: sys/hashes/sha256.c [hunk 5]
```diff
@@ -259,3 +259,62 @@ unsigned char *sha256(const unsigned char *d, size_t n, unsigned char *md)
 
     return md;
 }
+
+unsigned char * hmac_sha256(const unsigned char *key,
```
Comment (OlegHahm) [inline sys/hashes/sha256.c]: `s/char * hmac/char *hmac/`

File: sys/hashes/sha256.c [hunk 6]
```diff
@@ -259,3 +259,62 @@ unsigned char *sha256(const unsigned char *d, size_t n, unsigned char *md)
...
+                            size_t key_length,
+                            const unsigned *message,
```
Comment (OlegHahm) [inline sys/hashes/sha256.c]: const unsigned char?

File: sys/hashes/sha256.c [hunk 7]
```diff
@@ -259,3 +259,62 @@ unsigned char *sha256(const unsigned char *d, size_t n, unsigned char *md)
...
+                            size_t message_length,
+                            unsigned char *result)
+{
+    unsigned char k[64];
```
Comment (OlegHahm) [inline sys/hashes/sha256.c]: Can you replace 64 by a macro constant? (Could also go into a separate PR if you like.)

**Comment by OlegHahm** [comment]
Apart from the minor comments, I'm fine with the PR.

File: sys/hashes/sha256.c (hunk 6 above)
Comment (BytesGalore) [inline sys/hashes/sha256.c]: done.

File: sys/hashes/sha256.c (hunk 7 above)
Comment (BytesGalore) [inline sys/hashes/sha256.c]: done. `SHA256_INTERNAL_BLOCK_SIZE`

**Comment by OlegHahm** [comment]
Great. ACK and please squash!

**Comment by BytesGalore** [comment]
squashed
